### PR TITLE
[Merged by Bors] - Eisenstein series are bounded at infinity

### PR DESCRIPTION
**Pull request by CBirkbeck**
We show the `IsBoundedAtImInfty`  property for Eisenstein series.

---

- [x] depends on: #12600 
- [x] depends on: #12601
- [x] depends on: #12602
- [x] depends on: #11013 
- [x] depends on: #10377 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+noncomputable section
+
+open ModularForm EisensteinSeries UpperHalfPlane Set Filter Function Complex Matrix
+  SlashInvariantForm
+
+open scoped Topology BigOperators Nat Classical MatrixGroups
+
+namespace EisensteinSeries
+
+lemma norm_eisSummand_Summable (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:37]: I think
```suggestion
lemma summable_norm_eisSummand (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
```
but definitely with lower case "s" on summable

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+/-- The absolute value of the restricted sum is less than the full sum of the absolute values. -/
+lemma abs_le_tsum_abs (N : ℕ) (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
+    Complex.abs (eisensteinSeries a k z) ≤ ∑' (x : Fin 2 → ℤ), Complex.abs (eisSummand k x z) := by
+  simp_rw [← Complex.norm_eq_abs, eisensteinSeries]
+  apply le_trans (norm_tsum_le_tsum_norm ((norm_eisSummand_Summable k hk z).subtype _))
+    (tsum_subtype_le (fun (x : Fin 2 → ℤ) ↦ ‖(eisSummand k x z)‖) _ (fun _ ↦ norm_nonneg _)
+      (norm_eisSummand_Summable k hk z))
+
+/-- Eisenstein seires are bounded at infinity. -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:56]:
```suggestion
/-- Eisenstein series are bounded at infinity. -/
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+
+import Mathlib.NumberTheory.ModularForms.EisensteinSeries.UniformConvergence
+import Mathlib.Analysis.Normed.Order.Lattice
+import Mathlib.Analysis.Complex.UpperHalfPlane.FunctionsBoundedAtInfty
+import Mathlib.NumberTheory.ModularForms.Identities
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:10]:
```suggestion
import Mathlib.Analysis.Complex.UpperHalfPlane.FunctionsBoundedAtInfty
import Mathlib.Analysis.Normed.Order.Lattice
import Mathlib.NumberTheory.ModularForms.EisensteinSeries.UniformConvergence
import Mathlib.NumberTheory.ModularForms.Identities
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+/-- The absolute value of the restricted sum is less than the full sum of the absolute values. -/
+lemma abs_le_tsum_abs (N : ℕ) (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
+    Complex.abs (eisensteinSeries a k z) ≤ ∑' (x : Fin 2 → ℤ), Complex.abs (eisSummand k x z) := by
+  simp_rw [← Complex.norm_eq_abs, eisensteinSeries]
+  apply le_trans (norm_tsum_le_tsum_norm ((norm_eisSummand_Summable k hk z).subtype _))
+    (tsum_subtype_le (fun (x : Fin 2 → ℤ) ↦ ‖(eisSummand k x z)‖) _ (fun _ ↦ norm_nonneg _)
+      (norm_eisSummand_Summable k hk z))
+
+/-- Eisenstein seires are bounded at infinity. -/
+theorem eisensteinSeries_SIF_IsBoundedAtImInfty {N : ℕ+} (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k)
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:57]:
```suggestion
theorem isBoundedAtImInfty_eisensteinSeries_SIF {N : ℕ+} (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k)
```
(I have doubts about `_SIF` as well, but that can wait)

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      have := (summable_one_div_norm_rpow hk').mul_left <| r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k)
+      simp_rw [← Int.cast_neg, Real.rpow_intCast] at this
+      exact this
+    · intro x
+      have hk0 : 0 ≤ (k : ℝ) := by norm_cast; omega
+      have := summand_bound_of_mem_verticalStrip (z := Z) hk0 x (A := N) (B := 2) (two_pos)
+        (verticalStrip_anti_right (N : ℕ) hz hn)
+      simp_rw [eisSummand, Fin.isValue, one_div, norm_eq_abs, map_inv₀, map_zpow₀, ge_iff_le,
+        ← inv_zpow, inv_zpow', ← Real.rpow_intCast, Int.cast_neg]
+      exact this
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:78]:
```suggestion
  simp_rw [UpperHalfPlane.bounded_mem, eisensteinSeries_SIF] at *
  refine ⟨∑'(x : Fin 2 → ℤ), r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k) * ‖x‖ ^ (-k), 2, ?_⟩
  intro z hz
  obtain ⟨n, hn⟩ := (ModularGroup_T_zpow_mem_verticalStrip z N.2)
  rw [eisensteinSeries_slash_apply, ← eisensteinSeries_SIF_apply,
    ← T_zpow_width_invariant N k n (eisensteinSeries_SIF (a ᵥ* A) k) z]
  let Z := (ModularGroup.T ^ (N * n)) • z
  apply le_trans (abs_le_tsum_abs N (a ᵥ* A) k hk Z)
  apply tsum_le_tsum _ (norm_eisSummand_Summable k hk _)
  · have hk' : (2 : ℝ) < k := by norm_cast
    have := (summable_one_div_norm_rpow hk').mul_left <| r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k)
    simp_rw [← Int.cast_neg, Real.rpow_intCast] at this
    exact this
  · intro x
    have hk0 : 0 ≤ (k : ℝ) := by norm_cast; omega
    have := summand_bound_of_mem_verticalStrip (z := Z) hk0 x (A := N) (B := 2) (two_pos)
      (verticalStrip_anti_right (N : ℕ) hz hn)
    simp_rw [eisSummand, Fin.isValue, one_div, norm_eq_abs, map_inv₀, map_zpow₀, ge_iff_le,
      ← inv_zpow, inv_zpow', ← Real.rpow_intCast, Int.cast_neg]
    exact this
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    let Z := (ModularGroup.T ^ (N * n)) • z
+    apply le_trans (abs_le_tsum_abs N (a ᵥ* A) k hk Z)
+    apply tsum_le_tsum _ (norm_eisSummand_Summable k hk _)
+    · have hk' : (2 : ℝ) < k := by norm_cast
+      have := (summable_one_div_norm_rpow hk').mul_left <| r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k)
+      simp_rw [← Int.cast_neg, Real.rpow_intCast] at this
+      exact this
+    · intro x
+      have hk0 : 0 ≤ (k : ℝ) := by norm_cast; omega
+      have := summand_bound_of_mem_verticalStrip (z := Z) hk0 x (A := N) (B := 2) (two_pos)
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:74]:
```suggestion
      have := summand_bound_of_mem_verticalStrip (z := Z) hk0 x (A := N) (B := 2) two_pos
```
Do you really need to supply all these implicit arguments?

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean (hunk 4 above)
Comment (CBirkbeck) [reply Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:57]: We've been using it elsewhere so I kept it for consistency, but we could remove them all if needed.

**Comment by CBirkbeck** [comment]
> Have you considered using `z ^ (-k)` rather than `1 / z ^ k` in the `eisSummand` definition? It seems to make the algebra a bit shorter: [diff](https://github.com/leanprover-community/mathlib4/compare/eisensteinSeries_bounded_at_infty..HM-eisensteinSeries_bounded_at_infty)

Oh I didnt think of that, I didn't know if would this significant a change, but I'm happy with it. Is there an easy way to merge your suggestions into the PR?

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+We can then, first observe that the slash action just changes our `a` to `(a ᵥ* A)` and
+we then use our bounds for Eisenstein series in these vertical strips to get the result.
+-/
+
+noncomputable section
+
+open ModularForm EisensteinSeries UpperHalfPlane Set Filter Function Complex Matrix
+  SlashInvariantForm
+
+open scoped Topology BigOperators Nat Classical MatrixGroups
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:33]: What happens if you don't open `Classical`?

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace EisensteinSeries
+
+lemma summable_norm_eisSummand (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:37]:
```suggestion
lemma summable_norm_eisSummand {k : ℤ} (hk : 3 ≤ k) (z : ℍ) :
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    Summable fun (x : Fin 2 → ℤ) ↦ ‖(eisSummand k x z)‖ := by
+  have hk' : (2 : ℝ) < k := by norm_cast
+  apply ((summable_one_div_norm_rpow hk').mul_left <| r z ^ (-k : ℝ)).of_nonneg_of_le
+    (fun _ => Complex.abs.nonneg _)
+  intro b
+  simp only [eisSummand, map_zpow₀]
+  have hk0 : 0 ≤ (k : ℝ) := by positivity
+  exact_mod_cast summand_bound z hk0 b
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:45]:
```suggestion
  exact_mod_cast summand_bound z (show 0 ≤ (k : ℝ) by positivity) b
```
If you like it.

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+theorem isBoundedAtImInfty_eisensteinSeries_SIF {N : ℕ+} (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k)
+    (A : SL(2, ℤ)) : IsBoundedAtImInfty ((eisensteinSeries_SIF a k).toFun ∣[k] A) := by
+  simp_rw [UpperHalfPlane.bounded_mem, eisensteinSeries_SIF] at *
+  refine ⟨∑'(x : Fin 2 → ℤ), r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k) * ‖x‖ ^ (-k), 2, ?_⟩
+  intro z hz
+  obtain ⟨n, hn⟩ := (ModularGroup_T_zpow_mem_verticalStrip z N.2)
+  rw [eisensteinSeries_slash_apply, ← eisensteinSeries_SIF_apply,
+    ← T_zpow_width_invariant N k n (eisensteinSeries_SIF (a ᵥ* A) k) z]
+  let Z := (ModularGroup.T ^ (N * n)) • z
+  apply le_trans (abs_le_tsum_abs N (a ᵥ* A) k hk Z)
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:65]:
```suggestion
  apply le_trans (abs_le_tsum_abs N (a ᵥ* A) k hk _)
```
I think Lean is smart enough to find `Z` by itself.

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  apply tsum_le_tsum _ (summable_norm_eisSummand k hk _)
+  · have hk' : (2 : ℝ) < k := by norm_cast
+    exact_mod_cast (summable_one_div_norm_rpow hk').mul_left <| r ⟨⟨N, 2⟩, Nat.ofNat_pos⟩ ^ (-k)
+  · intro x
+    have hk0 : 0 ≤ (k : ℝ) := by positivity
+    simp_rw [eisSummand, norm_zpow]
+    exact_mod_cast
+      summand_bound_of_mem_verticalStrip hk0 x two_pos (verticalStrip_anti_right N hz hn)
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:73]:
```suggestion
    exact_mod_cast
      summand_bound_of_mem_verticalStrip (lt_trans two_pos hk').le x two_pos
      (verticalStrip_anti_right N hz hn)
```
(you have to move `hk'` before the last apply to make this working). It seems a bit an overkill to use `positivity` when you already know `2 < k`.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+/-- The absolute value of the restricted sum is less than the full sum of the absolute values. -/
+lemma abs_le_tsum_abs (N : ℕ) (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k) (z : ℍ) :
+    Complex.abs (eisensteinSeries a k z) ≤ ∑' (x : Fin 2 → ℤ), Complex.abs (eisSummand k x z) := by
+  simp_rw [← Complex.norm_eq_abs, eisensteinSeries]
+  apply le_trans (norm_tsum_le_tsum_norm ((summable_norm_eisSummand k hk z).subtype _))
+    (tsum_subtype_le (fun (x : Fin 2 → ℤ) ↦ ‖(eisSummand k x z)‖) _ (fun _ ↦ norm_nonneg _)
+      (summable_norm_eisSummand k hk z))
+
+/-- Eisenstein series are bounded at infinity. -/
+theorem isBoundedAtImInfty_eisensteinSeries_SIF {N : ℕ+} (a : Fin 2 → ZMod N) (k : ℤ) (hk : 3 ≤ k)
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:56]:
```suggestion
theorem isBoundedAtImInfty_eisensteinSeries_SIF {N : ℕ+} (a : Fin 2 → ZMod N) {k : ℤ} (hk : 3 ≤ k)
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean (hunk 7 above)
Comment (CBirkbeck) [reply Mathlib/NumberTheory/ModularForms/EisensteinSeries/IsBoundedAtImInfty.lean:33]: nothing, so I've removed it!

**Pull request closed without merging**